src/server/pmix_server_ops.c

```diff
@@ -557,47 +584,70 @@ static pmix_status_t _collect_data(pmix_server_trkr_t *trk,
                     rc = PMIX_ERR_NOT_FOUND;
                     PMIX_ERROR_LOG(rc);
                     PMIX_DESTRUCT(&cb);
-                    PMIX_DESTRUCT(&pbkt);
+                    PMIX_DESTRUCT(&rank_blobs);
                     goto cleanup;
                 }
                 rel_rank += pcs.rank;
 
                 /* pack the relative rank */
-                PMIX_CONSTRUCT(&pbkt, pmix_buffer_t);
-                PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, &pbkt,
+                pbkt = PMIX_NEW(pmix_buffer_t);
+                PMIX_CONSTRUCT(pbkt, pmix_buffer_t);
+                PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, pbkt,
                                  &rel_rank, 1, PMIX_PROC_RANK);
                 if (PMIX_SUCCESS != rc) {
                     PMIX_ERROR_LOG(rc);
                     PMIX_DESTRUCT(&cb);
-                    PMIX_DESTRUCT(&pbkt);
+                    PMIX_DESTRUCT(&rank_blobs);
+                    PMIX_RELEASE(pbkt);
                     goto cleanup;
                 }
                 /* pack the returned kval's */
                 PMIX_LIST_FOREACH(kv, &cb.kvs, pmix_kval_t) {
-                    PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, &pbkt, kv, 1, PMIX_KVAL);
-                    if (PMIX_SUCCESS != rc) {
+                    rc = pmix_gds_base_modex_pack_kval(&kmap, pbkt, kv);
+                    if (rc != PMIX_SUCCESS) {
                         PMIX_ERROR_LOG(rc);
                         PMIX_DESTRUCT(&cb);
-                        PMIX_DESTRUCT(&pbkt);
+                        PMIX_DESTRUCT(&rank_blobs);
+                        PMIX_RELEASE(pbkt);
                         goto cleanup;
                     }
                 }
-                /* extract the blob */
-                PMIX_UNLOAD_BUFFER(&pbkt, bo.bytes, bo.size);
-                PMIX_DESTRUCT(&pbkt);
-                /* pack the returned blob */
-                PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, &bucket,
-                                 &bo, 1, PMIX_BYTE_OBJECT);
-                PMIX_BYTE_OBJECT_DESTRUCT(&bo);
-                if (PMIX_SUCCESS != rc) {
-                    PMIX_ERROR_LOG(rc);
-                    PMIX_DESTRUCT(&cb);
-                    goto cleanup;
-                }
+
+                /* add part of the process modex to the list */
+                blob = PMIX_NEW(rank_blob_t);
+                blob->buf = pbkt;
+                pmix_list_append(&rank_blobs, &blob->super);
+                pbkt = NULL;
             }
             PMIX_DESTRUCT(&cb);
         }
+        /* pack node part of modex to `bucket` */
+        /* pack the key names map for the remote server can
+         * use it to match key names by index */
+        size = pmix_value_array_get_size(&kmap);
+        if (0 < size) {
+            PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, &bucket,
+                             &size, 1, PMIX_SIZE);
+            PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, &bucket,
```

Comment:
I think this is risky - the value_array class doesn't initialize its memory. You might be getting lucky and were given a memory location with zero's in it, but that is highly system dependent. If you weren't, then the array will not be NULL terminated and this will crash